crdb/retry.go

```diff
@@ -0,0 +1,284 @@
+		tryCount++
+		// Any negative value (including NoRetries) means fail immediately
+		if l.RetryLimit < UnlimitedRetries {
+			return 0, newMaxRetriesExceededError(err, 0)
+		}
+		// UnlimitedRetries (0) means retry indefinitely, so skip the limit check
+		// Any positive value enforces the retry limit
+		if l.RetryLimit > UnlimitedRetries && tryCount > l.RetryLimit {
+			return 0, newMaxRetriesExceededError(err, l.RetryLimit)
+		}
+		return l.Delay, nil
+	}
+}
+
+// ExpBackoffRetryPolicy implements RetryPolicy using an exponential backoff strategy
+// where delays double with each retry attempt, with an optional maximum delay cap.
+//
+// The delay between retries doubles with each attempt, starting from BaseDelay:
+//   - Retry 1: BaseDelay
+//   - Retry 2: BaseDelay * 2
+//   - Retry 3: BaseDelay * 4
+//   - Retry N: BaseDelay * 2^(N-1)
+//
+// If MaxDelay is set (> 0), the delay is capped at that value once reached.
+// This prevents excessive wait times during high retry counts and provides a
+// predictable upper bound for backoff duration.
+//
+// The policy will retry up to RetryLimit times. When the limit is exceeded or
+// if the delay calculation overflows without a MaxDelay set, it returns a
+// MaxRetriesExceededError.
+//
+// Example usage with capped exponential backoff:
+//
+//	policy := &ExpBackoffRetryPolicy{
+//	    RetryLimit: 10,
+//	    BaseDelay:  100 * time.Millisecond,
+//	    MaxDelay:   5 * time.Second,
+//	}
+//	ctx := crdb.WithRetryPolicy(context.Background(), policy)
+//	err := crdb.ExecuteTx(ctx, db, nil, func(tx *sql.Tx) error {
+//	    // transaction logic that may encounter retryable errors
+//	    return tx.ExecContext(ctx, "UPDATE ...")
+//	})
+//
+// This configuration produces delays: 100ms, 200ms, 400ms, 800ms, 1.6s, 3.2s,
+// then stays at 5s for all subsequent retries.
+//
+// Example usage with unbounded exponential backoff:
+//
+//	policy := &ExpBackoffRetryPolicy{
+//	    RetryLimit: 5,
+//	    BaseDelay:  1 * time.Second,
+//	    MaxDelay:   0,  // no cap
+//	}
+//
+// This configuration produces delays: 1s, 2s, 4s, 8s, 16s.
+// Note: Setting MaxDelay to 0 means no cap, but be aware that delay overflow
+// will cause the policy to fail early.
+type ExpBackoffRetryPolicy struct {
+	// RetryLimit is the maximum number of retries allowed. After this many
+	// retries, a MaxRetriesExceededError is returned.
+	RetryLimit int
+
+	// BaseDelay is the initial delay before the first retry. Each subsequent
+	// retry doubles this value: delay = BaseDelay * 2^(attempt-1).
+	BaseDelay time.Duration
+
+	// MaxDelay is the maximum delay cap. If > 0, delays are capped at this
+	// value once reached. If 0, delays grow unbounded (until overflow, which
+	// causes early termination).
+	MaxDelay time.Duration
+}
+
+// NewRetry implements RetryPolicy.
+func (l *ExpBackoffRetryPolicy) NewRetry() RetryFunc {
+	tryCount := 0
+	return func(err error) (time.Duration, error) {
+		tryCount++
+		if tryCount > l.RetryLimit {
```

Comment:
I think we should align the behavior of `RetryLimit == 0` for `ExpBackoffRetryPolicy` with `LimitBackoffRetryPolicy`. 

What do you think about makeing `0` mean "infinite retries" for both (which is what `LimitBackoffRetryPolicy` does currently)?

Reply:
Makes sense. Done.